plugins/derive/g_derive.mlg

```diff
@@ -20,10 +20,20 @@ DECLARE PLUGIN "coq-core.plugins.derive"
 
 let classify_derive_command _ = Vernacextend.(VtStartProof (Doesn'tGuaranteeOpacity,[]))
 
+let _ = Pcoq.(set_keyword_state (CLexer.add_keyword_tok (get_keyword_state ()) (Tok.PKEYWORD "SuchThat")))
```

Comment:
I'm not sure this works properly with backtracking.

Reply:
Good question. What is actually the status of backtracking with `VERNAC COMMAND EXTEND`, and the status of backtracking with plugins. If I try to backtrack over a `Require Import Derive` in a UI it seems to work but I don't know why.

Reply:
VERNAC COMMAND EXTEND declares an undoable grammar extension
cf around egramml extend_vernac_command_grammar ~undoable

Reply:
When I backtrack over `Require Import Derive.` (in CoqIDE), I can't anymore parse `Derive ...`. If I redo, it can parse again `Derive` and `SuchThat`. So, it seems that something is undone and redone.

Reply:
Not sure what coqide is doing, but in proofgeneral / coqtop
~~~coq
Require Derive.

Derive foo SuchThat (foo = 0) As baz.
~~~
go to end, go back before Require, go to end again says `Syntax error: ':=' or ':' or [name] expected after [name] (in [closed_binder]).`

also coqc and coqide on
~~~coq
Module M.

Require Derive.

End M.

Derive foo SuchThat (foo = 0) As baz.
~~~

Reply:
Indeed, backtracking with "go to cursor" did not work. I followed what is done in `VERNAC COMMAND EXTEND` and added an `Mltop.add_init_function`. It still is a bit of a hack, relying on a self-reference to the plugin name, but more explicit I think that declaring a keyword by side effect with `GRAMMAR EXTEND`. Is it ok to you, or should we still try to find better?

Reply:
The example
```
Module M.
Require Derive.
End M.
Derive foo SuchThat (foo = 0) As baz.
```
I added an example.